File: 46-permutations/46-permutations.py
class Solution:
    def permute(self, nums):
        # swapping algo
        res = []
        perm = []
        def backtrack(first):
            if first == len(nums) - 1:
                res.append(nums[:])
            for i in range(first, len(nums)):
                nums[first], nums[i] = nums[i], nums[first]
                backtrack(first + 1)
                nums[first], nums[i] = nums[i], nums[first]
        backtrack(0)
        return res
#     def permute(self, nums: List[int]) -> List[List[int]]:
#         res = []
#         perm = []
#         def backtrack():
#             if len(perm) == len(nums):
#                 res.append(perm.copy())
#                 return
            
#             for i in range(len(nums)):
#                 if nums[i] not in perm:
#                     perm.append(nums[i])
#                     backtrack()
#                     perm.pop()
#         backtrack()
#         return res

File: 46-permutations/46-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> perms = {};
        genPerms(0, nums, perms);
        return perms;
    }
    
    void genPerms(int i, vector<int>& curr, vector<vector<int>>& perms) {
        if(i == curr.size()){
            perms.push_back(curr);
            return;
        }
        
        for(int j = i; j < curr.size(); ++j){
            swap(curr[i], curr[j]);
            genPerms(i + 1, curr, perms);
            swap(curr[j], curr[i]);
        }
    }
};